File: src/client/java/io/github/ricciow/ChatManager.java
package io.github.ricciow;

import io.github.ricciow.config.PridgeConfig;
import io.github.ricciow.format.FormatManager;
import io.github.ricciow.format.FormatResult;
import io.github.ricciow.format.SpecialFunctions;
import io.github.ricciow.util.ColorCode;
import io.github.ricciow.util.TextParser;
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.hud.ChatHud;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;

import java.awt.*;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ChatManager {

    private final FormatManager FormatManager;

    private String incompleteMessage = "";
    private final String splitChar = "➩";
    private final PridgeConfig CONFIG;

    private static final Pattern GUILD_CHAT_PATTERN = Pattern.compile("^Guild > (.*?): (.*)$");
    private static final Pattern STATUS_CHAT_PATTERN = Pattern.compile("^Guild > (.*?) (joined|left)\\.$");
    private static final Pattern PRIVATE_CHAT_PATTERN = Pattern.compile("^From (.*?): (.*)$");

    private ChatHud CHAT_HUD;

    public ChatManager(FormatManager FormatManager) {
        this.FormatManager = FormatManager;

        CONFIG = PridgeClient.getConfig();
    }

    public void register() {
        ClientReceiveMessageEvents.ALLOW_GAME.register(this::onReceiveChatMessage);
    }

    private void sendMessage(Text message) {
        if(CHAT_HUD == null) {
            CHAT_HUD = MinecraftClient.getInstance().inGameHud.getChatHud();
        }
        CHAT_HUD.addMessage(message);
    }

    public boolean onReceiveChatMessage(Text message, boolean overlay) {
        if(CONFIG.developerCategory.enabled) {
            if(overlay) return true;

            String cleanRawMessage = Formatting.strip(message.getString());
            if (cleanRawMessage == null) {
                return true;
            }

            String wordFilters = CONFIG.filtersCategory.rawFilter;
            if(!Objects.equals(wordFilters, "")) {
                Pattern filterRegex = Pattern.compile(wordFilters);
                Matcher filterMatches = filterRegex.matcher(cleanRawMessage);

                if(filterMatches.find()) {
                    if(CONFIG.filtersCategory.placeholder) {
                        sendMessage(TextParser.parseHoverable("&c&lA message has been filtered.", message));
                    }
                    return false;
                }
            }

            if(CONFIG.soundsCategory.enabled) {
                PridgeClient.SOUND_PLAYER.checkForSounds(cleanRawMessage);
            }

            //Guild Chat Message handling
            Matcher guildMatcher = GUILD_CHAT_PATTERN.matcher(cleanRawMessage);
            if (guildMatcher.matches()) {
                return onReceiveGuildMessage(message, guildMatcher);
            }

            //Join/Leave handling
            Matcher statusMatcher = STATUS_CHAT_PATTERN.matcher(cleanRawMessage);
            if (statusMatcher.matches()) {
                return onReceiveStatusMessage(message, statusMatcher);
            }

            //Direct Message handling
            Matcher privateMatcher = PRIVATE_CHAT_PATTERN.matcher(cleanRawMessage);
            if(privateMatcher.matches()) {
                return onReceivePrivateMessage(message, privateMatcher);
            }
        }
        return true;
    }

    private boolean onReceiveGuildMessage(Text originalMessage, Matcher guildMatcher) {
        String userInfo = guildMatcher.group(1).trim();
        String chatContent = guildMatcher.group(2).trim();

        chatContent = chatContent.replaceAll("<@\\S+>", "").trim();

        Optional<String> userOptional = Arrays.stream(userInfo.split(" "))
                .filter(part -> !part.startsWith("[") && !part.endsWith("]"))
                .findFirst();

        // Only proceed if we successfully found a username.
        if (userOptional.isPresent()) {
            if(userOptional.get().equalsIgnoreCase(CONFIG.botCategory.ign)) {
                return onReceiveBotMessage(originalMessage, userInfo, chatContent);
            }
            return onReceivePlayerMessage(originalMessage, userInfo, chatContent);
        }
        return true;
    }

    private boolean onReceiveBotMessage(Text originalMessage, String userInfo, String chatContent) {
        boolean startsWithSplit = chatContent.startsWith(splitChar);
        boolean endsWithSplit = chatContent.endsWith(splitChar);
        boolean isBuffering = !incompleteMessage.isEmpty();

        String finalContent = null;

        if (isBuffering) {
            if (startsWithSplit) {
                String middlePart = chatContent.substring(1);
                if (endsWithSplit) {
                    incompleteMessage += middlePart.substring(0, middlePart.length() - 1);
                    return false;
                } else {
                    finalContent = incompleteMessage + middlePart;
                    incompleteMessage = "";
                }
            } else {
                incompleteMessage = "";
                finalContent = chatContent;
            }
        } else {
            if (endsWithSplit) {
                incompleteMessage = chatContent.substring(0, chatContent.length() - 1);
                return false;
            } else {
                finalContent = chatContent;
            }
        }

        if (finalContent != null) {
            FormatResult formattedContent = FormatManager.formatText(finalContent);
            if (formattedContent != null) {
                sendMessage(formattedContent.getText());
                return false;
            }
        }

        return true;
    }

    private boolean onReceivePlayerMessage(Text originalMessage, String userInfo, String chatContent) {
        if(CONFIG.guildCategory.modify_normal_guild_messages) {
            sendMessage(TextParser.parse(originalMessage.getString().replace("§2Guild >", CONFIG.guildCategory.name)));
            return false;
        }
        return true;
    }

    private boolean onReceiveStatusMessage(Text originalMessage, Matcher matcher) {
        if(CONFIG.guildCategory.modify_join_leave) {

            String guildTag = CONFIG.guildCategory.modify_normal_guild_messages ? CONFIG.guildCategory.name : "&2Guild >";
            ColorCode colorCode = matcher.group(2).equals("left") ? ColorCode.RED : ColorCode.GREEN;

            String finalResult = guildTag + " " + ColorCode.GOLD.getCode() + originalMessage.getString().split(" ")[2] + " " + colorCode.getCode() + matcher.group(2);
            sendMessage(TextParser.parse(finalResult));
            return false;
        }
        return true;
    }

    private boolean onReceivePrivateMessage(Text originalMessage, Matcher matcher) {
        if(CONFIG.guildCategory.thanks_for_the_boop) {
            String userInfo = matcher.group(1).trim();

            Optional<String> userOptional = Arrays.stream(userInfo.split(" "))
                    .filter(part -> !part.startsWith("[") && !part.endsWith("]"))
                    .findFirst();

            if(userOptional.isPresent()) {
                MinecraftClient client = MinecraftClient.getInstance();
                if (client.player != null) {
                    if (matcher.group(2).equals("Boop!")) {
                        client.player.networkHandler.sendChatCommand("gc Thanks for the Boop, " + userOptional.get() + "!");
                    } else if (matcher.group(2).equals("Boo!")) {
                        client.player.networkHandler.sendChatCommand("gc AAH! You scared me, " + userOptional.get() + "!");
                    }
                }
            }

        }
        return true;
    }
}

File: src/main/kotlin/io/github/ricciow/ChatManager.kt
package io.github.ricciow

import io.github.ricciow.Pridge.CONFIG_I
import io.github.ricciow.Pridge.mc
import io.github.ricciow.format.FormatManager
import io.github.ricciow.sounds.DynamicSoundPlayer
import io.github.ricciow.util.*
import io.github.ricciow.util.TextParser.parse
import io.github.ricciow.util.TextParser.parseHoverable
import net.fabricmc.fabric.api.client.message.v1.ClientReceiveMessageEvents
import net.minecraft.util.Formatting
import java.util.regex.Pattern

interface ChatResult {
    fun handle(): Boolean
}

data class BotMessageResult(
    val message: String
) : ChatResult {
    override fun handle(): Boolean {
        if (handlePartial(message)) {
            PridgeLogger.dev("Partial message was handled: $message")
        }

        var finalContent: String
        if (multiMessageBuffered) {
            finalContent = multiMessage
            multiMessageBuffered = false
            multiMessage = ""
        } else if (multiMessage.isNotEmpty()) {
            return false
        } else {
            finalContent = message
        }

        val formattedContent = FormatManager.formatText(finalContent)
        PridgeLogger.dev("Message was formatted to: $formattedContent")
        ChatUtils.sendMessage(formattedContent.getText())
        return false
    }

    private fun handlePartial(message: String): Boolean {
        if (!message.contains(SPLIT_CHAR)) {
            return false
        } else {
            multiMessage += message.replace(SPLIT_CHAR, "")
            multiMessageBuffered = !message.endsWith(SPLIT_CHAR)
            return true
        }
    }

    companion object {
        private var multiMessageBuffered = false
        private var multiMessage = ""
        private const val SPLIT_CHAR = "➩"
    }
}

data class PlayerMessageResult(
    val user: String,
    val original: String,
    val message: String
) : ChatResult {
    override fun handle(): Boolean {
        if (CONFIG_I.guildCategory.modifyNormalGuildMessages) {
            val parsable = TextParser.replaceMinecraftColorCodes(original)
            ChatUtils.sendMessage(parse(parsable.replace("&2Guild >", CONFIG_I.guildCategory.name)))
            return false
        }
        return true
    }
}

enum class StatusAction(val color: ColorCode) {
    JOINED(ColorCode.GREEN), LEFT(ColorCode.RED);

    fun actionString(): String {
        return when (this) {
            JOINED -> "${color.getMcCode()}joined"
            LEFT -> "${color.getMcCode()}left"
        }
    }
}

data class StatusChatResult(
    val user: String,
    val action: StatusAction
) : ChatResult {
    override fun handle(): Boolean {
        if (CONFIG_I.guildCategory.modifyJoinLeave) {
            val guildTag =
                if (CONFIG_I.guildCategory.modifyNormalGuildMessages) CONFIG_I.guildCategory.name else "&2Guild >"
            ChatUtils.sendMessage(
                parse("$guildTag ${ColorCode.GOLD.getMcCode()}${user} ${action.actionString()}")
            )
            return false
        }
        return true
    }
}

data class PrivateChatResult(
    val user: String,
    val message: String
) : ChatResult {
    override fun handle(): Boolean {
        if (CONFIG_I.guildCategory.thanksForTheBoop) {
            val boopboo = BoopBooType.fromString(message)
            if (boopboo != null) {
                return BoopBooResult(user, boopboo).handle()
            }
        }

        return true
    }
}

enum class BoopBooType {
    BOOP, BOO;

    companion object {
        fun fromString(type: String): BoopBooType? {
            return when (type) {
                "Boop!" -> BOOP
                "Boo!" -> BOO
                else -> null
            }
        }
    }
}

data class BoopBooResult(
    val user: String,
    val type: BoopBooType
) : ChatResult {
    override fun handle(): Boolean {
        mc.player.chatHypixel(
            ChatType.GUILD, when (type) {
                BoopBooType.BOOP -> CONFIG_I.guildCategory.boopReplyMessage.replace("{user}", user)
                BoopBooType.BOO -> CONFIG_I.guildCategory.booReplyMessage.replace("{user}", user)
            }
        )
        return true
    }
}

object ChatManager {
    private val GUILD_CHAT_PATTERN = Regex("""^Guild > (?:\[(?<hypixelRank>[\w+]+)] )?(?<username>\w{2,16})(?: \[(?<guildRank>[\w+]+)])?: (?<content>.+?)(?: <@\w+>)?$""")
    private val STATUS_CHAT_PATTERN = Regex("""^Guild > (?<username>\w{2,16}) (?<action>joined|left)\.$""")
    private val PRIVATE_CHAT_PATTERN = Regex("""^From (?:\[(?<hypixelRank>[\w+]+)] )?(?<name>\w{2,16}): (?<content>.+)$""")

    fun initialize() {
        ClientReceiveMessageEvents.ALLOW_GAME.register { message, overlay ->
            if (!CONFIG_I.developerCategory.enabled) return@register true
            if (overlay) return@register true

            PridgeLogger.dev("Received message: $message")

            val cleanRawMessage = Formatting.strip(message.string) ?: return@register true

            if (checkWordFilters(cleanRawMessage)) return@register false

            checkSounds(cleanRawMessage)

            // Add handlers for different chat types
            val chatResult =
                GUILD_CHAT_PATTERN.matchEntire(cleanRawMessage)?.let { handleGuildChat(it, message.string) } ?: // Breaks clean chain :wah:
                STATUS_CHAT_PATTERN.matchEntire(cleanRawMessage)?.let(::handleStatusChat) ?:
                PRIVATE_CHAT_PATTERN.matchEntire(cleanRawMessage)?.let(::handlePrivateChat) ?:
                return@register true

            chatResult.handle()
        }
    }

    private fun handleGuildChat(matchResult: MatchResult, original: String): ChatResult? {
        val username = matchResult.groups["username"]!!.value
        val content = matchResult.groups["content"]!!.value

        // Only proceed if we successfully found a username.
        if (username.equals(CONFIG_I.botCategory.ign, ignoreCase = true)) {
            return BotMessageResult(content)
        }
        return PlayerMessageResult(username, original, content)
    }

    private fun handleStatusChat(matchResult: MatchResult): StatusChatResult? {
        val username = matchResult.groups["username"]!!.value
        val actionString = matchResult.groups["action"]!!.value
        val action = when (actionString) {
            "joined" -> StatusAction.JOINED
            "left" -> StatusAction.LEFT
            else -> return null // Invalid action, do not handle
        }
        return StatusChatResult(username, action)
    }

    private fun handlePrivateChat(matchResult: MatchResult): PrivateChatResult? {
        val name = matchResult.groups["name"]!!.value.trim()
        val content = matchResult.groups["content"]!!.value.trim()
        return PrivateChatResult(name, content)
    }

    private fun checkWordFilters(message: String): Boolean {
        val wordFilters = try {
            val filterStr = CONFIG_I.filtersCategory.rawFilter
            if (filterStr.isBlank()) null
            Regex(filterStr)
        } catch (e: Exception) {
            PridgeLogger.dev("Failed to compile word filters regex: ${e.message}")
            null
        }

        if (wordFilters != null) {
            if (wordFilters.matches(message)) {
                if (CONFIG_I.filtersCategory.placeholder) {
                    ChatUtils.sendMessage(parseHoverable("&c&lA message has been filtered.", message))
                }
                return true
            }
        }
        return false
    }

    private fun checkSounds(message: String) {
        if (CONFIG_I.soundsCategory.enabled) {
            DynamicSoundPlayer.playSoundIfMessageContains(message)
        }
    }
}